extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/scan/ScanQueryKit.java

```diff
@@ -178,6 +183,11 @@ public QueryDefinition makeQueryDefinition(
       }
     }
 
+    // If window has an empty over, we want a single worker to process entire data for window function evaluation.
+    if (windowHasEmptyOver) {
+      scanShuffleSpec = MixShuffleSpec.instance();
```

Comment:
We also do not respect the resultShuffleSpecFactory argument. This means that if we are adding a segment generator stage etc, we might not have correctly sized segments. Could we also add a test with an ingestion and see if the segments being generated are respecting any rowsPerSegment parameters?

Reply:
> This means that if we are adding a segment generator stage etc, we might not have correctly sized segments.

Can you please guide me on how to validate this? What kind of query to try?

> Could we also add a test with an ingestion and see if the segments being generated are respecting any rowsPerSegment parameters?

Do you mean ingestion with a window function?

Reply:
Yes. Do an ingestion into a window function which generates 3 rows of results. Set rowsPerSegment to 1 and see how many partitions are being created.

Reply:
As mentioned on the [other comment](https://github.com/apache/druid/pull/16804#discussion_r1696410451), MSQ ingestion with window functions is broken right now. Adding more details here though.

I tried this query:
```sql
REPLACE INTO "test_1" OVERWRITE ALL
WITH "ext" AS (
  select __time, countryName, cityName, channel
  from "drill_wikipedia"
  where countryName in ('Guatemala')
  group by __time, countryName, cityName, channel 
  order by countryName, cityName, channel
)
SELECT
  "__time",
  "countryName",
  "channel",
  row_number() over (partition by cityName order by countryName, cityName, channel) as c1,
  count(channel) over (partition by cityName order by countryName, cityName, channel) as c2
FROM "ext"
PARTITIONED BY DAY
```

with the following query context:
```json
{
  "maxNumTasks": 2,
  "enableWindowing": true,
  "rowsPerSegment": 1
}
```

It fails because of NPE, but it doesn't even go to the ScanQueryKit. The following stages are created for the above query:
<img width="1227" alt="image" src="https://github.com/user-attachments/assets/dfccd261-9e49-4706-87ba-076c943a7ef6">

I also tried changing `maxNumTasks`, but it runs into the same issue.

So, I feel that my changes in ScanQueryKit should be unrelated? This NPE is certainly something we need to fix, but it's also not because of this PR's changes.

Thoughts?

Reply:
For your reference, the stacktrace of the NPE is:
```java
java.lang.NullPointerException
	at org.apache.druid.frame.key.RowKeyReader.read(RowKeyReader.java:82)
	at org.apache.druid.msq.exec.ControllerImpl.getBucketDateTime(ControllerImpl.java:2269)
	at org.apache.druid.msq.exec.ControllerImpl.getPartitionsByBucket(ControllerImpl.java:1052)
	at org.apache.druid.msq.exec.ControllerImpl.generateSegmentIdsWithShardSpecsForReplace(ControllerImpl.java:1122)
	at org.apache.druid.msq.exec.ControllerImpl.generateSegmentIdsWithShardSpecs(ControllerImpl.java:923)
	at org.apache.druid.msq.exec.ControllerImpl.access$1400(ControllerImpl.java:239)
	at org.apache.druid.msq.exec.ControllerImpl$RunQueryUntilDone.populateSegmentsToGenerate(ControllerImpl.java:2791)
	at org.apache.druid.msq.exec.ControllerImpl$RunQueryUntilDone.startStages(ControllerImpl.java:2736)
	at org.apache.druid.msq.exec.ControllerImpl$RunQueryUntilDone.run(ControllerImpl.java:2512)
	at org.apache.druid.msq.exec.ControllerImpl$RunQueryUntilDone.access$000(ControllerImpl.java:2461)
	at org.apache.druid.msq.exec.ControllerImpl.runInternal(ControllerImpl.java:391)
	at org.apache.druid.msq.exec.ControllerImpl.run(ControllerImpl.java:339)
	at org.apache.druid.msq.indexing.MSQControllerTask.runTask(MSQControllerTask.java:290)
	at org.apache.druid.indexing.common.task.AbstractTask.run(AbstractTask.java:179)
	at org.apache.druid.indexing.overlord.ThreadingTaskRunner$1.call(ThreadingTaskRunner.java:229)
	at org.apache.druid.indexing.overlord.ThreadingTaskRunner$1.call(ThreadingTaskRunner.java:154)
	at com.google.common.util.concurrent.TrustedListenableFutureTask$TrustedFutureInterruptibleTask.runInterruptibly(TrustedListenableFutureTask.java:131)
	at com.google.common.util.concurrent.InterruptibleTask.run(InterruptibleTask.java:75)
	at com.google.common.util.concurrent.TrustedListenableFutureTask.run(TrustedListenableFutureTask.java:82)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)
```

Reply:
If ingestion is working is not fully relevant, I meant that the changes to how we create shuffleSpec might be incorrect. You can also test this by running an async query, with rowsPerPage parameter, and see if the correct number of pages are created. This codeflow would be hit by both querying and ingestion, so we would want at least the querying flow to be right.

Also, I think we should fix the NPE in this PR if the fix is small and straightforward.

Reply:
> Also, I think we should fix the NPE in this PR if the fix is small and straightforward.

I haven't dug into the ingestion issues yet, so I'm unsure of the scope there. But since it's unrelated to this PR, I'd prefer taking it up separately in future.

> You can also test this by running an async query, with rowsPerPage parameter, and see if the correct number of pages are created

I tried this. Summarising what I am getting:

Query:
```sql
select
"countryName",
"cityName",
row_number() over (partition by cityName order by countryName, cityName, channel) as c1,
count(channel) over (partition by cityName order by countryName, cityName, channel) as c2,
"channel"
from "drill_wikipedia"
where countryName in ('Guatemala')
group by countryName, cityName, channel 
order by countryName, cityName, channel
```

## Query context 1

```json
{
  "maxNumTasks": 4,
  "enableWindowing": true,
  "rowsPerSegment": 1,
  "selectDestination": "durableStorage",
  "durableShuffleStorage": true
}
```

<img width="1637" alt="image" src="https://github.com/user-attachments/assets/c0a9a0e3-557f-48cd-9631-d48f4ff82cd3">


## Query context 2 (added `rowsPerPage: 1`)

```json
{
  "maxNumTasks": 4,
  "enableWindowing": true,
  "rowsPerPage": 1,
  "rowsPerSegment": 1,
  "selectDestination": "durableStorage",
  "durableShuffleStorage": true
}
```

<img width="1632" alt="image" src="https://github.com/user-attachments/assets/4e6b0ea9-559f-4bf1-872f-7b4019a7b610">

Reply:
@adarshsanjeev I've updated the PR as per offline discussion. Will close this comment as it's not relevant in the latest changes.